simplesat/dependency_solver.py

```diff
@@ -342,26 +338,33 @@ def _create_rules_and_initialize_policy(self, request):
             if len(providers) == 0:
                 raise NoPackageFound(requirement, str(requirement))
 
-            if job.kind == JobType.update:
+            if job.kind == JobType.hard_update:
                 # An update request *must* install the latest package version
                 def key(package):
                     return (package.version, package in installed_repository)
                 providers = [max(providers, key=key)]
+            elif job.kind == JobType.soft_update:
+                soft_update_packages.update((pkg for pkg in providers
+                                            if pkg in installed_repository))
 
-            requirement_ids = [pool.package_id(p) for p in providers]
-            self._policy.add_requirements(requirement_ids)
-            all_requirement_ids.extend(requirement_ids)
+            all_requirement_ids.extend([pool.package_id(p) for p in providers])
 
         installed_package_ids = collections.OrderedDict()
         for package in installed_repository:
             package_id = pool.package_id(package)
             installed_package_ids[package_id] = package
 
+        # Prefer the installed versions of all packages
+        policy = InstalledFirstPolicy(
+            pool, installed_repository,
+            ignore_installed_packages=soft_update_packages)
+        policy.add_requirements(all_requirement_ids)
```

Comment:
we are now `passing all_requirements_ids` instead of `[pool.package_id(p) for p in providers]` to `policy.add_requirements`. What are the consequences (if any) ?